fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/TriState.java

```diff
@@ -16,16 +16,197 @@
+
+/**
+ * Represents a boolean value which can be true, false or refer to a default value.
+ */
 public enum TriState {
+	/**
+	 * Represents the boolean value of {@code false}.
+	 */
 	FALSE,
+	/**
+	 * Represents a value that refers to a "default" value, often as a fallback.
+	 */
 	DEFAULT,
+	/**
+	 * Represents the boolean value of {@code true}.
+	 */
 	TRUE;
 
-	public static TriState of(boolean b) {
-		return b ? TRUE : FALSE;
+	/**
+	 * Gets the corresponding tri-state from a boolean value.
+	 *
+	 * @param bool the boolean value
+	 * @return {@link TriState#TRUE} or {@link TriState#FALSE} depending on the value of the boolean.
+	 */
+	public static TriState of(boolean bool) {
+		return bool ? TRUE : FALSE;
+	}
+
+	/**
+	 * Gets a tri-state from a nullable boxed boolean.
+	 *
+	 * @param bool the boolean value
+	 * @return {@link TriState#DEFAULT} if {@code null}.
+	 * Otherwise {@link TriState#TRUE} or {@link TriState#FALSE} depending on the value of the boolean.
+	 */
+	public static TriState of(/* @Nullable */ Boolean bool) {
+		if (bool == null) {
+			return DEFAULT;
+		}
+
+		return TriState.of(bool.booleanValue());
 	}
 
+	/**
+	 * Gets the value of the tri-state.
+	 *
+	 * @return true if the tri-state is {@link TriState#TRUE}.
+	 * Otherwise false.
+	 */
 	public boolean get() {
 		return this == TRUE;
 	}
+
+	/**
+	 * Gets the value of the tri-state as a boxed, nullable boolean.
+	 *
+	 * @return {@code null} if {@link TriState#DEFAULT}.
+	 * Otherwise {@code true} if {@link TriState#TRUE} or {@code false} if {@link TriState#FALSE}.
+	 */
+	/* @Nullable */
+	public Boolean getBoxed() {
+		if (this == TriState.DEFAULT) {
+			return null;
+		}
+
+		return this.get();
+	}
+
+	/**
+	 * Gets the value of this tri-state.
+	 * If the value is {@link TriState#DEFAULT} then use the supplied value.
+	 *
+	 * @param value the value to fallback to
+	 * @return the value of the tri-state or the supplied value if {@link TriState#DEFAULT}.
+	 */
+	public boolean orElse(boolean value) {
+		return this == TriState.DEFAULT ? value : this.get();
+	}
+
+	/**
+	 * Gets the value of this tri-state.
+	 * If the value is {@link TriState#DEFAULT} then use the supplied value.
+	 *
+	 * @param supplier the supplier used to get the value to fallback to
+	 * @return the value of the tri-state or the value of the supplier if the tri-state is {@link TriState#DEFAULT}.
+	 */
+	public boolean orElseGet(BooleanSupplier supplier) {
+		return this == TriState.DEFAULT ? supplier.getAsBoolean() : this.get();
+	}
+
+	/**
+	 * Maps the value of an optional if this tri-state is {@link TriState#TRUE}.
+	 *
+	 * @param mapper the mapper to use
+	 * @param <T> the type of object supplied by the mapper
+	 * @return an optional containing the value if {@link TriState#TRUE} and the value returned by the mapper is non-null, otherwise {@link Optional#empty()}.
+	 */
+	public <T> Optional<T> mapTrue(Supplier<T> mapper) {
+		Objects.requireNonNull(mapper);
```

Comment:
Could be good to make a private `mapEq` or `mapValue` common method that takes in a Supplier<T> and a TriState because `mapTrue`, `mapFalse`, and `mapDefault` all use the same logic except they use different things to compare to.

Take for example
```java
private <T> Optional<T> mapEq(TriState comparison, Supplier<T> mapper) {
    Objects.requireNonNull(mapper);

    if (this == comparison) {
        return Optional.ofNullable(mapper.get());
    }

    return Optional.empty();
}

public <T> Optional<T> mapTrue(Supplier<T> mapper) {
    return this.mapEq(TriState.TRUE, mapper);
}

public <T> Optional<T> mapFalse(Supplier<T> mapper) {
    return this.mapEq(TriState.FALSE, mapper);
}

public <T> Optional<T> mapDefault(Supplier<T> mapper) {
    return this.mapEq(TriState.DEFAULT, mapper);
}
```